docs/proposals/ingester-unregister-flexibility.md

```diff
@@ -0,0 +1,129 @@
+* Give Mimir operators more flexibility to choose between availability and consistency.
+
+## Proposal
+
+A new `/ingester/prepare-unregister` HTTP endpoint is added.
+This endpoint can be used to control whether the ingester should unregister from the ring the next time it is stopped.
+
+The endpoint supports three HTTP methods: `GET`, `PUT` and `DELETE`.
+
+When called with the `GET` method, the endpoint returns the current unregister state in a JSON response body with a `200` status code:
+```json
+{"unregister": true}
+```
+
+When called with the `PUT` method, the endpoint accepts a request body:
+```json
+{"unregister": false}
+```
+The endpoint sets `i.lifecycler.SetUnregisterOnShutdown()` to the value passed in the request body. After doing so, it returns the current lifecycler unregister value with a `200` status code:
+```json
+{"unregister": false}
+```
+
+Using a request body supports both of these use cases:
+1. Disabling unregister for an ingester that has it enabled by default.
+2. Enabling unregister for an ingester that has it disabled by default.
+
+When called with the `DELETE` method, the endpoint sets the ingester's unregister state to what was passed via the `unregister_on_shutdown` configuration option. After doing so, it returns the current unregister value with a `200` status code:
+```json
+{"unregister": false}
+```
+The `DELETE` method is considered to "delete" any override, and can be used to ensure that the ingester's unregister state is set to the value that was set on ingester startup.
+
+All three behaviours of the endpoint are idempotent.
+
+### Example usage
+
+The `/ingester/prepare-unregister` endpoint allows operators to run ingesters in a default mode of _not_ leaving the
+ring on shutdown.
+
+Most cloud platforms offer an "eviction notice" prior to node eviction. A helper service can be set up to react to
+the eviction notice by updating evicted ingesters to unregister from the ring.
+
+Concrete example:
+1. Ingesters `a-1` and `a-7` run on node `zone1-snkq`. Ingesters `b-4` and `b-9` run on node `zone2-iqmd`.
+2. An eviction notice is sent for nodes `zone1-snkq` and `zone2-iqmd` at the same time.
+3. Helper service picks up the eviction notice, figures out which ingesters run on the two nodes, and invokes the
+   `/ingester/prepare-unregister` endpoint to make each ingester leave the ring when evicted.
+4. The ingesters are evicted and leave the ring.
+5. Because the ingesters leave the ring, each ingester zone is left with exclusively `ACTIVE` ingesters, and availability
+   is maintained.
+
+**Pros**:
+* Availability is maintained even during multi-zone evictions.
+
+**Cons**:
+* Queries may return incomplete results as samples held by the evicted ingesters are temporarily inaccessible.
+  The samples become accessible again when the ingesters join the ring after having replayed their WAL.
+* Ownership of evicted series (i.e., series whose owner was evicted) is split between two ingesters. This is likely
```

Comment:
To confirm my understanding, this is a result of resharding, right?

Reply:
> this is a result of resharding, right?

Yes. Consider this example:
1. Ingester `a-1` owns series `up{app="database"}`.
2. `a-1` ingests a couple of samples from series `up{app="database"}`.
3. `a-1` is evicted.
4. As soon as `a-1` is evicted, ownership of `up{app="database"}` is implicitly passed to `a-5`.
5. While `a-1` is restarting, `a-5` ingests samples from the `up{app="database"}` series.
6. Ownership of series `up{app="database"}` remains with `a-5` even after `a-1` has joined the ring once again[^1].
7. Ownership of series `up{app="database"}` is now split since both `a-1` and `a-5` hold samples for the series.

[^1]: Unless tokens are persisted on disk via `-ingester.ring.tokens-file-path`, in which case ownership of the series _will_ be passed back to `a-1`.

Reply:
Right, thank you. I find this aspect of the proposal quite worrisome, to be honest, but whether it is a problem or not depends on your specifics.

(I will "unresolve" this conversation, as it has context that others may find useful.)

Reply:
@pstibrany Is there anything in particular with split series ownership that concerns you?

If my understanding is correct, split ownership can already happen if ingesters are configured (in the standard way) to leave the ring on shutdown. I suppose the endpoint makes it easier to trigger a state of split ownership.

Reply:
> split ownership can already happen if ingesters are configured (in the standard way) to leave the ring on shutdown

That's correct. However, we don't recommend to leave ring on shutdown (e.g. we don't do it in our Jsonnet and Helm). The only reason why the default is still `-ingester.ring.unregister-on-shutdown=true` is because it's easier for the first time Mimir operator (at the beginning it's difficult to understand the implications of not leaving the ring on shutdown).

Reply:
> @pstibrany Is there anything in particular with split series ownership that concerns you?

Resharding of the series and related increased usage of resources (memory, cpu) on other ingesters. If ingesters leave ring frequently, surviving ingesters may accumulate lot of series in memory. It's something to keep in mind when planning for capacity. You save cost thanks to using spot-instances, but you may need to plan for larger instances instead. As you mention, it depends on how often ingester evictions happen.